R/zchunk_LA100.IEA_downscale_ctry.R

```diff
@@ -0,0 +1,308 @@
+#' @param command API command to execute
+#' @param ... other optional parameters, depending on command
+#' @return Depends on \code{command}: either a vector of required inputs,
+#' a vector of output names, or (if \code{command} is "MAKE") all
+#' the generated outputs: \code{L100.IEA_en_bal_ctry_hist}. The corresponding file in the
+#' original data system was \code{LA100.IEA_downscale_ctry.R} (energy level1).
+#' @details Combine OECD and non-OECD data; perform upfront adjustments for other Africa, Turkey, and South
+#' Africa; split out and handle the 1990 split of Yugoslavia and USSR, back-projecting individual country
+#' values based on 1990 shares; use population to downscale IEA composite regions (Other Africa, Other
+#' non-OECD Americas, Other Asia) to individual countries; filter out countries without data in any year.
+#' @note We build from the raw (and proprietary) \code{en_OCED} and \code{en_nonOECD} files if
+#' they are available; if not, this chunk reads in a pre-generated \code{L100.IEA_en_bal_ctry_hist}
+#' file and returns that. (In other words, our output is an optional input.)
+#' @importFrom assertthat assert_that
+#' @importFrom dplyr filter mutate select
+#' @importFrom tidyr gather spread
+#' @author BBL May 2017
+module_energy_LA100.IEA_downscale_ctry <- function(command, ...) {
+  if(command == driver.DECLARE_INPUTS) {
+    return(c("L100.Pop_thous_ctry_Yh",
+             OPTIONAL_FILE = "energy/en_OECD",
+             OPTIONAL_FILE = "energy/en_nonOECD",
+             OPTIONAL_FILE = "energy/L100.IEA_en_bal_ctry_hist",
+             FILE = "energy/mappings/IEA_product_downscaling",
+             FILE = "energy/mappings/IEA_ctry"))
+  } else if(command == driver.DECLARE_OUTPUTS) {
+    return(c("L100.IEA_en_bal_ctry_hist"))
+  } else if(command == driver.MAKE) {
+
+    `1990` <- `1990_share` <- `1990_sum` <- COUNTRY <- EcYield_kgm2_hi <- EcYield_kgm2_lo <- FLOW <-
+      GCAM_commodity <- GCAM_region_ID <- GLU <- IEAcomposite <- Irr_Rfd <- LC_bm2_hi <-
+      LC_bm2_lo <- PRODUCT <- composite_population <- iso <- landshare_hi <- landshare_lo <-
+      level <- optional <- value <- year <- yield <- yieldmult_hi <- yieldmult_lo <- NULL # silence package check notes
+
+    all_data <- list(...)[[1]]
+
+    # Load required inputs
+    L100.Pop_thous_ctry_Yh <- get_data(all_data, "L100.Pop_thous_ctry_Yh")
+    en_OECD <- get_data(all_data, "energy/en_OECD")
+    en_nonOECD <- get_data(all_data, "energy/en_nonOECD")
+    IEA_product_downscaling <- get_data(all_data, "energy/mappings/IEA_product_downscaling")
+    IEA_ctry <- get_data(all_data, "energy/mappings/IEA_ctry")
+    L100.IEA_en_bal_ctry_hist_preprocessed <- get_data(all_data, "energy/L100.IEA_en_bal_ctry_hist")
+
+    # If the (proprietary) raw IEA datasets are available, go through the full computations below
+    # If not, use the pre-saved summary file (i.e., the output of this chunk!) assuming it's available
+    if(!is.na(en_nonOECD) && !is.na(en_OECD)) {
```

Comment:
`is.na` returns a vector value, and so it is not really the right thing to use in this test.  It will generate a warning when the relevant files are present (or worse, if the first entry in the table happens to be `NA`, it will read as true.  Generally, you want to represent a missing _object_ (as opposed to a missing _value_) with `NULL` and test it with `is.null`.